src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,328 @@
+
+                var shard = TcpConnection.ShardedContext.GetShardedDatabaseName(i);
+
+                var shardWorker = new ShardedSubscriptionWorker(_options, shard, re);
+
+                var worker = new SubscriptionShardHolder
+                {
+                    Worker = shardWorker,
+                    PullingTask = shardWorker.RunInternal(CancellationTokenSource.Token),
+                    RequestExecutor = re
+                };
+
+                _shardWorkers.Add(shard, worker);
+            }
+
+            await WriteJsonAsync(new DynamicJsonValue
+            {
+                [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.ConnectionStatus),
+                [nameof(SubscriptionConnectionServerMessage.Status)] = nameof(SubscriptionConnectionServerMessage.ConnectionStatus.Accepted)
+            });
+        }
+
+        private async Task MaintainConnectionWithClientWorker()
+        {
+            var exceptions = new Dictionary<string, Exception>();
+            var shardsToReconnect = new List<string>();
+            var heartbeatSp = Stopwatch.StartNew();
+            var reconnectSp = Stopwatch.StartNew();
+            var didBatch = false;
+            while (CancellationTokenSource.IsCancellationRequested == false)
+            {
+                exceptions.Clear();
+                shardsToReconnect.Clear();
+                if (didBatch == false)
+                    await Task.Delay(500);
+
+                if (heartbeatSp.ElapsedMilliseconds >= 3000)
+                {
+                    await SendHeartBeat("Waited for 3000ms for batch from shard workers");
+                    heartbeatSp.Restart();
+                }
+                didBatch = false;
+                var stopping = false;
+                foreach ((string shard, SubscriptionShardHolder shardHolder) in _shardWorkers)
+                {
+                    CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                    if (shardHolder.PullingTask.IsFaulted == false)
+                    {
+                        ShardedSubscriptionWorker.PublishedBatch batch = shardHolder.Worker.PublishedBatchItem;
+                        if (batch == null)
+                            continue;
+
+                        didBatch = true;
+                        // Send to client
+                        try
+                        {
+                            await WriteBatchToClientAndAck(batch, shard);
+                            batch.SendBatchToClientTcs.SetResult();
+                        }
+                        catch (Exception e)
+                        {
+                            // need to fail the shard subscription worker
+                            batch.SendBatchToClientTcs.TrySetException(e);
+                            throw;
+                        }
+                        await batch.ConfirmFromShardSubscriptionConnectionTcs.Task;
+
```

Comment:
what are we sending to the client in the meantime?

Reply:
added comment